lib/asset/AssetGenerator.js

```diff
@@ -44,11 +46,12 @@ class AssetGenerator extends Generator {
 	generate(module, { chunkGraph, runtimeTemplate, runtimeRequirements, type }) {
 		switch (type) {
 			case "asset":
+				module.buildInfo.assetFilename = this.filename;
 				return module.originalSource();
 			default: {
 				runtimeRequirements.add(RuntimeGlobals.module);
 
-				if (module.buildInfo.dataUrl) {
+				if (module.buildInfo.dataUrl && !this.filename) {
```

Comment:
I don't think it makes sense to disable DataUrl when filename is provided. `type: "asset/resource"` already disables DataUrl, providing a `filename` to `type: "asset"` should still automatically select DataUrl vs resource, and use filename for the resource.